### PR TITLE
[Bug-Fix] TorchDataLoaderAdapter corrupts dict and single-tensor batches on non-Torch backends

### LINKED ISSUE #23080: [Bug] TorchDataLoaderAdapter corrupts dict and single-tensor batches on non-Torch backends
In `keras/src/trainers/data_adapters/torch_data_loader_adapter.py`, `get_numpy_iterator()` converts torch tensor leaves with `tree.map_structure(...)` but then wraps the whole result in `tuple(...)`. This changes the root structure:

1. A dict batch becomes a tuple of dict keys
2. A single tensor batch becomes a tuple over the leading batch dim
3. The TF path can fail while creating `output_signature` because it also wraps `get_tensor_spec(...)` in tuple
(This only affects the non-Torch paths)

Minimal repro:

```python
import os
os.environ.setdefault("KERAS_BACKEND", "tensorflow")

import torch
from keras.src.trainers.data_adapters.torch_data_loader_adapter import (
    TorchDataLoaderAdapter,
)

class DictDataset(torch.utils.data.Dataset):
    def __len__(self):
        return 2

    def __getitem__(self, idx):
        return {
            "x": torch.tensor([idx, idx + 1], dtype=torch.float32),
            "y": torch.tensor(idx, dtype=torch.float32),
        }

class TensorOnlyDataset(torch.utils.data.Dataset):
    def __len__(self):
        return 4

    def __getitem__(self, idx):
        return torch.tensor([idx, idx + 1, idx + 2], dtype=torch.float32)

for name, dataset in [
    ("dict", DictDataset()),
    ("single_tensor", TensorOnlyDataset()),
]:
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=2)
    adapter = TorchDataLoaderAdapter(dataloader)

    converted = next(adapter.get_numpy_iterator())
    print(name, type(converted), repr(converted))

    try:
        tfds = TorchDataLoaderAdapter(dataloader).get_tf_dataset()
        print(name, tfds.element_spec)
    except Exception as e:
        print(name, type(e).__name__, e)

```

Output:
```
dict <class 'tuple'> ('x', 'y')
dict TypeError `output_signature` must contain objects that are subclass of `tf.TypeSpec` but found <class 'str'> which is not.

single_tensor <class 'tuple'> (array([0., 1., 2.], dtype=float32), array([1., 2., 3.], dtype=float32))
single_tensor TypeError 'TensorSpec' object is not iterable
```

Actually dict should remain dict and single tensors should remain single tensors.

Cause:
The root-level tuple(...) calls at:
keras/src/trainers/data_adapters/torch_data_loader_adapter.py (line 37 and 57)
iterate over whatever structure is returned

Will put up a PR with fix. If this was intentional for TF compatibility, should it be limited to list containers so mappings and tensor leaves keep their structure?

### PR DESCRIPTION
**Pull request by ChiragSW**
## Description
Fixes: #23080 
Removed the force change of the dict and single tensor batches to tuple(...). Now tree utilities preserve the batch structure. Added a tf specific normalization so now nested lists to tuples without converting dicts or tensors accidentally. Added regression tests for the same.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
